graphql/parser.go

```diff
@@ -368,6 +368,52 @@ func Parse(source string, vars map[string]interface{}) (*Query, error) {
 		return nil, NewClientError("must have a single query")
 	}
 
+	// Parse variable definitions, default values, etc.
+	var defaultedVars map[string]interface{}
+	for _, variableDefinition := range queryDefinition.VariableDefinitions {
+		name := variableDefinition.Variable.Name.Value
+
+		if _, ok := variableDefinition.Type.(*ast.NonNull); ok {
+			if variableDefinition.DefaultValue != nil {
+				return nil, NewClientError("required varaible cannot provide a default value: $%s", name)
+			}
+
+			if vars[name] == nil {
+				return nil, NewClientError("required variable not provided: $%s", name)
+			}
+
+			continue
+		}
+
+		if variableDefinition.DefaultValue != nil {
+			// Ignore default if the value exists.
+			if vars[name] != nil {
+				continue
+			}
+
+			// Lazily initialize defaultedVars if needed.
+			if defaultedVars == nil {
+				defaultedVars = make(map[string]interface{})
+				for k, v := range vars {
+					defaultedVars[k] = v
+				}
+			}
+
+			// TODO: properly implement coerceValue.
+			// See: https://github.com/graphql/graphql-js/blob/17a0bfd5292f39cafe4eec5b3bd0e22514243b68/src/execution/values.js#L84
```

Comment:
note that without this, we are not properly checking that the default value can be validly used for the value.